core/referral/engine.go

```diff
@@ -49,6 +66,77 @@ type Engine struct {
 	// proposal to apply at the start of the next epoch.
 	// It's `nil` is there is none.
 	newProgram *types.ReferralProgram
+
+	// TODO: snapshot this
+	// keep track of all referral sets
+	// referral set ID -> Referral set
+	sets map[string]*types.ReferralSet
+
+	// NO need for snapshot, dynamically computed
+	// map of referrer to set ID
+	referrers map[string]string
+
+	// NO need for snapshot, dynamically computed
+	// map of referrees to set ID
+	referrees map[string]string
+}
+
+func (e *Engine) Exists(setID string) bool {
+	_, ok := e.sets[setID]
+	return ok
+}
+
+func (e *Engine) CreateReferralSet(ctx context.Context, party string, set *commandspb.CreateReferralSet, deterministicID string) error {
+	if _, ok := e.referrers[party]; ok {
+		return ErrIsAlreadyAReferrer(party)
+	}
+
+	now := e.timeSvc.GetTimeNow()
+
+	e.sets[deterministicID] = &types.ReferralSet{
+		ID:        deterministicID,
+		CreatedAt: now,
+		UpdatedAt: now,
+		Referrer: &types.Membership{
+			PartyID:  types.PartyID(party),
+			JoinedAt: now,
+			// TODO:
+			// not sure we need number of epoch
```

Comment:
Yes we do, but not as is, it should be a `StartAtEpoch`. There is an issue for that

Reply:
alright, cool. let's fix it in that PR?